feat: Make AWS credentials optional with fallback to default credential chain

annsway opened the pull request:
### Description

- Make the AWS credential fields optional in our pipeline YAML and code schema.

- Wrap the [WithCredentialsProvider(...)](https://github.com/conduitio-labs/conduit-connector-dynamodb/blob/add468bb2242ff9809e1b20506633c297352eb1c/source.go#L89) logic so it only applies when those fields are non-empty, otherwise falling back to[ LoadDefaultConfig’s default chain](https://docs.aws.amazon.com/sdk-for-go/v2/developer-guide/configure-gosdk.html#specifying-credentials).

- Update the README/examples to mark those fields “optional” and add a note about IRSA (or env/shared-file) as the recommended production approach.

This change allows the connector to authenticate with AWS using either static credentials or AWS default config chain, providing more flexibility in how the connector accesses DynamoDB resources.

### Testing

Tested locally by dropping the aws.accessKeyId and aws.secretAccessKey fields from our connector and instead exporting them in my ~/.zshrc. The [AWS SDK for Go v2](https://github.com/aws/aws-sdk-go-v2) picked them up automatically and authenticated successfully.

<img width="1271" alt="Screenshot 2025-05-14 at 1 48 19 PM" src="https://github.com/user-attachments/assets/5bc84c7f-0fb1-42fa-bb8e-3f050d119355" />

<details>
<summary> Passed integration test
</summary>

```
% make test-integration
docker compose -f test/docker-compose.yml up --quiet-pull -d --wait
[+] Running 2/2
 ✔ Network test_default         Created                                                                                               0.0s 
 ✔ Container test-localstack-1  Healthy                                                                                              11.7s 
go test  -v -race ./...; ret=$?; \
                docker compose -f test/docker-compose.yml down -v; \
                exit $ret
=== RUN   TestSource_SuccessfulSnapshot
Table conduit-dynamodb-source-test-144e8011-2bcf-410d-a295-88bbbfe39894 is now active.
--- PASS: TestSource_SuccessfulSnapshot (8.34s)
=== RUN   TestSource_SnapshotRestart
Table conduit-dynamodb-source-test-1a301971-d1ff-4510-bb33-7a5fbcb1ce08 is now active.
--- PASS: TestSource_SnapshotRestart (5.15s)
=== RUN   TestSource_EmptyTable
Table conduit-dynamodb-source-test-2fdf1d41-f3bc-4c9c-bb19-9b9c28b0b3a7 is now active.
--- PASS: TestSource_EmptyTable (11.18s)
=== RUN   TestSource_NonExistentTable
Table conduit-dynamodb-source-test-7e7446a5-6be7-40ef-81da-6915a6f2dc99 is now active.
--- PASS: TestSource_NonExistentTable (0.04s)
=== RUN   TestSource_CDC
Table conduit-dynamodb-source-test-2e8ff876-a39f-407c-bedc-94683cb5f2a7 is now active.
--- PASS: TestSource_CDC (11.16s)
PASS
ok      github.com/conduitio-labs/conduit-connector-dynamodb    37.546s
?       github.com/conduitio-labs/conduit-connector-dynamodb/cmd/connector      [no test files]
?       github.com/conduitio-labs/conduit-connector-dynamodb/iterator   [no test files]
=== RUN   TestParseSDKPosition
=== RUN   TestParseSDKPosition/valid_position
=== RUN   TestParseSDKPosition/unknown_iterator_type
=== RUN   TestParseSDKPosition/nil_position
--- PASS: TestParseSDKPosition (0.00s)
    --- PASS: TestParseSDKPosition/valid_position (0.00s)
    --- PASS: TestParseSDKPosition/unknown_iterator_type (0.00s)
    --- PASS: TestParseSDKPosition/nil_position (0.00s)
PASS
ok      github.com/conduitio-labs/conduit-connector-dynamodb/position   1.773s
[+] Running 2/2
 ✔ Container test-localstack-1  Removed                                                                                               2.5s 
 ✔ Network test_default         Removed        
```

</details>

### Quick checks:

- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-dynamodb/pulls) for the same update/change.
- [ ] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

lovromazgon commented on source.go:
```diff
@@ -82,15 +85,43 @@ func (s *Source) Config() sdk.SourceConfig {
+        opts = append(opts,
+            config.WithAssumeRoleCredentialOptions(func(o *stscreds.AssumeRoleOptions) {
+                o.RoleARN = s.config.AWSRoleARN
+                o.RoleSessionName = "conduit-dynamodb-connector"
```
I'd make this also configurable through the config. Also, I'd change the default to `conduit-connector-dynamodb`, just so it exactly matches the connector name.

lovromazgon commented on source.go:
```diff
@@ -82,15 +85,43 @@ func (s *Source) Config() sdk.SourceConfig {
+    }
+
+    // If the user supplied static creds, register them first
+    if s.config.AWSAccessKeyID != "" && s.config.AWSSecretAccessKey != "" {
```
I'm thinking if this condition should rather be:

```suggestion
    if s.config.AWSAccessKeyID != "" || s.config.AWSSecretAccessKey != "" || s.config.AWSSessionToken != "" {
```

Since all three values are used in that credentials provider.

Thinking out loud - does it ever make sense to provide a role _and_ the static credentials? Or should we disallow this to prevent a misconfiguration?

annsway replied:
The PR has been updated. The AWSAccessKeyID and AWSSecretAccessKey will be needed for static credentials. The IAM role config will be handled implicitly using AWS default authentication chain. See details in my PR description. 
```
// Only use static credentials if both access key and secret are provided
	if s.config.AWSAccessKeyID != "" && s.config.AWSSecretAccessKey != ""
```

lovromazgon commented on README.md:
```diff
@@ -42,17 +42,17 @@ pipelines:
           # Type: string
-          # Required: yes
+          # Required: no
           aws.accessKeyId: ""
```
This part of the readme is automatically generated, you can install the needed tools using `make install-tools` and then `make generate`.

annsway replied:
Updated the comments in both README.md and connector.yaml using `make generate`.

lovromazgon commented on connector.yaml:
```diff
@@ -3,17 +3,15 @@ specification:
     parameters:
       - name: aws.accessKeyId
-        description: AWS access key id.
+        description: AWS access key id. Required if using static credentials. If using IAM role (aws.roleArn), this can be left empty.
```
Same goes for this part of the `connector.yaml` file, it will be automatically updated when executing `make generate`. The data is taken from the comments and struct tags on the `SourceConfig` struct.

annsway replied:
Updated the comments in both README.md and connector.yaml using make generate.

lovromazgon reviewed (commented):
Thanks for the contribution @annsway!

annsway committed:
- Make the AWS credential fields optional in our pipeline YAML and code schema.

- Wrap the WithCredentialsProvider(...) logic so it only applies when those fields are non-empty, otherwise falling back to LoadDefaultConfig’s default chain.

- Update the README/examples to mark those fields “optional” and add a note about IRSA (or env/shared-file) as the recommended production approach.

lyuboxa commented on source.go:
```diff
@@ -84,10 +90,28 @@ func (s *Source) Config() sdk.SourceConfig {
+	}
+
+	// Only use static credentials if both access key and secret are provided
+	if s.config.AWSAccessKeyID != "" && s.config.AWSSecretAccessKey != "" {
```
@annsway Can you move this part to a separate function either on `Config` or in `Source` to return `[]func(*config.LoadOptions)` 

E.g. 

```golang
func (c Config) AWSLoadOpts() []func(*config.LoadOptions) {

}
```
then here
```golang
config.LoadDefaultConfig(ctx, s.config.AWSLoadOptions()...)
```

This should resolve the linter.

annsway replied:
Thanks for the info. Updated.

annsway committed:
encapsulate all AWS config logic in one function
